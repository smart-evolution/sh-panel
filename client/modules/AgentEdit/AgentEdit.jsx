// @flow
import React from 'react';
import { withRouter } from 'react-router';
import * as agentsConstants from 'client/models/agents/constants';
import * as agentTypes from 'client/models/agents/types';
import * as agentQueries from 'client/models/agents/queries';
import * as agentConfigTypes from 'client/models/agentConfigs/types';

type Props = {|
  agent: agentTypes.Agent,
  agentConfig: agentConfigTypes.AgentConfig,
  commitConfig: (agentTypes.AgentID, agentConfigTypes.AgentConfig) => void,
  updateProperty: (agentTypes.AgentID, string, string) => void,
|};

class AgentEdit extends React.Component<Props> {
  constructor(props) {
    super(props);
    (this: any).updateTemperature = this.updateTemperature.bind(this);
    (this: any).updateName = this.updateName.bind(this);
    (this: any).updateConfig = this.updateConfig.bind(this);
  }

  updateConfig() {
    const { agent, agentConfig, commitConfig } = this.props;
    commitConfig(agent.id, agentConfig);
  }

  updateTemperature(e) {
    const value = e.target.value;
    const { agent, updateProperty } = this.props;
    updateProperty(agent.id, 'temperature', value);
  }

  updateName(e) {
    const value = e.target.value;
    const { agent, updateProperty } = this.props;
    updateProperty(agent.id, 'name', value);
  }

  render() {
    const { agentConfig, agent } = this.props;

    const rawType = agentQueries.getNoVersionedType(agent);

    const temperatureAdjustment = (
      <div className="c-input c-input__full">
        <div className="c-input__label">Temperature modifier</div>
        <input
          className="c-input__field"
          value={agentConfig.temperature || ''}
          onChange={this.updateTemperature}
        />
      </div>
    );

    return (
      <div>
        <div className="c-input c-input__full">
          <div className="c-input__label">Name</div>
          <input
            className="c-input__field"
            value={agentConfig.name || ''}
            onChange={this.updateName}
          />
        </div>
        {rawType === agentsConstants.Type1 && temperatureAdjustment}
        <button
          className="c-btn c-btn--full c-btn--accept"
          onClick={this.updateConfig}
        >
          UPDATE
        </button>
      </div>
    );
  }
}

export default withRouter(AgentEdit);
