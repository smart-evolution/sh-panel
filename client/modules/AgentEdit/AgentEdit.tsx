import React from 'react';
import { withRouter } from 'react-router';
import * as agentsConstants from 'client/models/agents/constants';
import type * as agentTypes from 'client/models/agents/types';
import * as agentQueries from 'client/models/agents/queries';
import type * as agentConfigTypes from 'client/models/agentConfigs/types';
import { Dialog, Button } from 'graphen';

type Props = {
  agent: agentTypes.Agent;
  agentConfig: agentConfigTypes.AgentConfig;
  commitConfig: (
    arg0: agentTypes.AgentID,
    arg1: agentConfigTypes.AgentConfig
  ) => void;
  updateProperty: (
    arg0: agentTypes.AgentID,
    arg1: string,
    arg2: string
  ) => void;
  removeAgent: (arg0: agentTypes.AgentID) => void;
};

type State = {
  isDialogShown: boolean;
};

class AgentEdit extends React.Component<Props, State> {
  constructor(props) {
    super(props);
    this.updateTemperature = this.updateTemperature.bind(this);
    this.updateName = this.updateName.bind(this);
    this.updateConfig = this.updateConfig.bind(this);
    this.removeAgent = this.removeAgent.bind(this);
    this.handleDialogToggle = this.handleDialogToggle.bind(this);

    this.state = {
      isDialogShown: false,
    };
  }

  handleDialogToggle(isDialogShown: boolean) {
    this.setState({
      isDialogShown,
    });
  }

  updateConfig() {
    const { agent, agentConfig, commitConfig } = this.props;
    commitConfig(agent.id, agentConfig);
  }

  updateTemperature(e) {
    const { value } = e.target;
    const { agent, updateProperty } = this.props;
    updateProperty(agent.id, 'temperature', value);
  }

  updateName(e) {
    const { value } = e.target;
    const { agent, updateProperty } = this.props;
    updateProperty(agent.id, 'name', value);
  }

  removeAgent() {
    const { agent, removeAgent } = this.props;
    removeAgent(agent.id);
  }

  render() {
    const { isDialogShown } = this.state;
    const { agentConfig, agent } = this.props;

    const rawType = agentQueries.getNoVersionedType(agent);

    const temperatureAdjustment = (
      <div className="gc-input gc-input--full">
        <div className="gc-input__label">Temperature modifier</div>
        <input
          className="gc-input__field"
          value={agentConfig.temperature || ''}
          onChange={this.updateTemperature}
        />
      </div>
    );

    return (
      <div className="gc-panel gc-panel--separator">
        <div className="gc-flex">
          <div className="gc-flex__item gm-spacing-rl gc-card gc-card--gradient gc-panel">
            <div className="gc-panel__title">Adjustments</div>
            <div className="gc-panel__content">
              <p>Adjust agent with custom settings.</p>
              <div className="gc-input gc-input--full">
                <div className="gc-input__label">Name</div>
                <input
                  className="gc-input__field"
                  value={agentConfig.name || ''}
                  onChange={this.updateName}
                />
              </div>
              {rawType === agentsConstants.Type1 && temperatureAdjustment}
            </div>
            <div className="gc-panel__footer">
              <button
                type="button"
                className="gc-btn gc-btn--full gc-btn--primary"
                onClick={this.updateConfig}
              >
                Update
              </button>
            </div>
          </div>
          <div className="gc-flex__item gc-card gc-card--gradient gc-panel">
            <div className="gc-panel__title">Remove agent</div>
            <div className="gc-panel__content">Remove agent permanently.</div>
            <div className="gc-panel__footer">
              <button
                type="button"
                className="tst-delete gc-btn gc-btn--full gc-btn--primary"
                onClick={() => this.handleDialogToggle(true)}
              >
                Delete
              </button>
            </div>
          </div>
        </div>
        {isDialogShown && (
          <Dialog>
            <article className="gc-panel">
              <header className="gc-panel__title">Remove confirmation</header>
              <div className="gc-panel__content">
                Are you sure you want to remove this agent?
              </div>
              <div className="gc-panel__footer">
                <Button
                  onClick={() => this.removeAgent()}
                  className="tst-delete-confirm gc-btn--danger"
                >
                  Delete
                </Button>{' '}
                <Button
                  onClick={() => this.handleDialogToggle(false)}
                  className="gc-btn--secondary"
                >
                  Cancel
                </Button>
              </div>
            </article>
          </Dialog>
        )}
      </div>
    );
  }
}

export default withRouter(AgentEdit);
