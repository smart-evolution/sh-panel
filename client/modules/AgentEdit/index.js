// @flow
import { connect } from 'react-redux';
import * as actions from 'client/models/agentConfigs/actions';
import * as agentConfigSelectors from 'client/models/agentConfigs/selectors';
import * as agentSelectors from 'client/models/agents/selectors';
import AgentEdit from './AgentEdit';

const mapStateToProps = (state, ownProps) => {
  const {
    match: { params },
  } = ownProps;
  const agentID = params.agent;
  const agentConfig =
    agentConfigSelectors.getAgentConfigById(state, agentID) || {};
  const agent = agentSelectors.getAgentById(state, agentID);

  return {
    timestamp: new Date(),
    agent,
    agentConfig,
  };
};

const mapDispatchToProps = dispatch => ({
  updateProperty: (agentID, key, value) => {
    dispatch(actions.updateProperty(agentID, key, value));
  },
  commitConfig: (agentID, config) => {
    dispatch(actions.commitAgentConfig(agentID, config));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AgentEdit);
