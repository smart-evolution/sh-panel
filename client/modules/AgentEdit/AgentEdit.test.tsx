import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter as Router } from 'react-router';
import AgentEdit from './AgentEdit';

const agent = {
  id: '90346453',
  name: 'Livingroom',
  data: {},
  type: 'type1-v0.3.3',
  ip: '192.168.1.10',
  isOnline: true,
};

const agentConfig = {
  id: '90346453',
  name: 'Livingroom',
  temperature: 0,
};

describe('AgentEdit', () => {
  it('should render correctly', () => {
    const component = mount(
      <Router keyLength={0}>
        <AgentEdit agent={agent} agentConfig={agentConfig} />
      </Router>
    );

    expect(component).toMatchSnapshot();
  });
});
