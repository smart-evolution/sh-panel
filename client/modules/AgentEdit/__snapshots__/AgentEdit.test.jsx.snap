// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgentEdit should render correctly 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(AgentEdit)
      agent={
        Object {
          "data": Object {},
          "id": "90346453",
          "ip": "192.168.1.10",
          "isOnline": true,
          "name": "Livingroom",
          "type": "type1-v0.3.3",
        }
      }
      agentConfig={
        Object {
          "id": "90346453",
          "name": "Livingroom",
          "temperature": 0,
        }
      }
    >
      <Route>
        <AgentEdit
          agent={
            Object {
              "data": Object {},
              "id": "90346453",
              "ip": "192.168.1.10",
              "isOnline": true,
              "name": "Livingroom",
              "type": "type1-v0.3.3",
            }
          }
          agentConfig={
            Object {
              "id": "90346453",
              "name": "Livingroom",
              "temperature": 0,
            }
          }
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <div>
            <div
              className="c-input c-input__full"
            >
              <div
                className="c-input__label"
              >
                Name
              </div>
              <input
                className="c-input__field"
                onChange={[Function]}
                value="Livingroom"
              />
            </div>
            <div
              className="c-input c-input__full"
            >
              <div
                className="c-input__label"
              >
                Temperature modifier
              </div>
              <input
                className="c-input__field"
                onChange={[Function]}
                value=""
              />
            </div>
            <button
              className="c-btn c-btn--full c-btn--accept"
              onClick={[Function]}
            >
              UPDATE
            </button>
          </div>
        </AgentEdit>
      </Route>
    </withRouter(AgentEdit)>
  </Router>
</MemoryRouter>
`;
